### ISSUE TITLE
Blob Metadata Not Working

### ISSUE THREAD

**Issue by rrmistry** [terraform-providers/terraform-provider-azurerm#4901]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Summary

Looks like blob [`metadata`](https://www.terraform.io/docs/providers/azurerm/r/storage_blob.html#metadata) attribute is not working anymore (or as expected)
![image](https://user-images.githubusercontent.com/25514145/69014210-eb228680-0955-11ea-9bbd-b90ee4ee7459.png)

### Terraform (and AzureRM Provider) Version

```
Terraform v0.12.15
+ provider.archive v1.3.0
+ provider.azurerm v1.36.0
```

### Affected Resource(s)

* `azurerm_storage_blob`

### Terraform Configuration Files

```hcl
provider "azurerm" {
  # whilst the `version` attribute is optional, we recommend pinning to a given version of the Provider
  version = "~> 1.3"

  # Ref: https://github.com/terraform-providers/terraform-provider-azurerm/issues/4440
  skip_provider_registration = "true"

  subscription_id = var.az_subscription_id
}

# A zip file to package Azure Functions App source
data "archive_file" "package" {
  type        = "zip"
  output_path = "${path.module}/azure.zip"
  source_dir  = "${path.module}/dist"
}

resource "azurerm_storage_account" "default" {
  name                     = var.project_name
  resource_group_name      = var.az_resource_group
  location                 = var.az_region
  account_tier             = "Standard"
  account_replication_type = "LRS"
  tags                     = var.az_tags
}

resource "azurerm_storage_container" "default" {
  name                  = var.project_name
  storage_account_name  = azurerm_storage_account.default.name
  container_access_type = "private"
}

resource "azurerm_storage_blob" "default" {
  name                   = "azure.zip"
  storage_account_name   = azurerm_storage_account.default.name
  storage_container_name = azurerm_storage_container.default.name
  type                   = "block"
  source                 = data.archive_file.package.output_path
  metadata = {
    # Force re-deploy if hash changes
    PACKAGE_HASH = data.archive_file.package.output_base64sha256
  }
}
```

### Debug Output

```
------
2019/11/17 16:13:11 [TRACE] Executing graph transform *terraform.OrphanResourceCountTransformer
2019/11/17 16:13:11 [TRACE] Completed graph transform *terraform.OrphanResourceCountTransformer (no changes)
2019/11/17 16:13:11 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2019/11/17 16:13:11 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2019/11/17 16:13:11 [TRACE] Executing graph transform *terraform.TargetsTransformer
2019/11/17 16:13:11 [TRACE] Completed graph transform *terraform.TargetsTransformer (no changes)
2019/11/17 16:13:11 [TRACE] Executing graph transform *terraform.ReferenceTransformer
2019/11/17 16:13:11 [DEBUG] ReferenceTransformer: "azurerm_storage_container.default" references: []
2019/11/17 16:13:11 [TRACE] Completed graph transform *terraform.ReferenceTransformer (no changes)
2019/11/17 16:13:11 [TRACE] Executing graph transform *terraform.RootTransformer
2019/11/17 16:13:11 [TRACE] Completed graph transform *terraform.RootTransformer (no changes)
2019/11/17 16:13:11 [TRACE] vertex "azurerm_storage_container.default": entering dynamic subgraph
2019/11/17 16:13:11 [TRACE] dag/walk: updating graph
2019/11/17 16:13:11 [TRACE] dag/walk: added new vertex: "azurerm_storage_container.default"
202019/11/17 16:13:11 [ERROR] <root>: eval: *terraform.EvalDiff, err: MetaData must start with letters or an underscores. Got "PACKAGE_HASH".
2019/11/17 16:13:11 [ERROR] <root>: eval: *terraform.EvalSequence, err: MetaData must start with letters or an underscores. Got "PACKAGE_HASH".
2019/11/17 16:13:11 [TRACE] [walkPlan] Exiting eval tree: azurerm_storage_blob.default
2019/11/17 16:13:11 [TRACE] vertex "azurerm_storage_blob.default": visit complete
2019/11/17 16:13:11 [TRACE] vertex "azurerm_storage_blob.default": dynamic subgraph encountered errors
2019/11/17 16:13:11 [TRACE] vertex "azurerm_storage_blob.default": visit complete
2019/11/17 16:13:11 [TRACE] dag/walk: upstream of "azurerm_function_app.default" errored, so skipping
2019/11/17 16:13:11 [TRACE] dag/walk: upstream of "provider.azurerm (close)" errored, so skipping
2019/11/17 16:13:11 [TRACE] dag/walk: upstream of "output.azurerm_function_app_id" errored, so skipping
2019/11/17 16:13:11 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2019/11/17 16:13:11 [TRACE] dag/walk: upstream of "root" errored, so skipping
2019/11/17 16:13:11 [INFO] backend/local: plan operation completed
2019/11/17 16:13:11 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info

Error: MetaData must start with letters or an underscores. Got "PACKAGE_HASH".2019-11-17T16:13:11.739-0500 [DEBUG] plugin: plugin process exited: path=XXXXXXXXX\.terraform\plugins\windows_amd64\terraform-provider-azurerm_v1.36.0_x4.exe pid=18828
2019-11-17T16:13:11.739-0500 [DEBUG] plugin: plugin exited


  on azurefunction.tf line 27, in resource "azurerm_storage_blob" "default":
  27: resource "azurerm_storage_blob" "default" {

```

### Panic Output

None

### Expected Behavior

Detect existing or new value for Azure Blob metadata (without any errors)

### Actual Behavior

Errors out when providing [`metadata`](https://www.terraform.io/docs/providers/azurerm/r/storage_blob.html#metadata) map

### Steps to Reproduce

1. `terraform plan`

### Important Factoids

None beyond what is provided above

### References

* #1129
* #1185
* #1400
* #3206

**Comment by rrmistry** [comment]
I am using [`metadata`](https://www.terraform.io/docs/providers/azurerm/r/storage_blob.html#metadata) attribute to force Terraform to re-deploy my `archive_file` to `azurerm_storage_blob`. I have a workaround using [`tags`](https://www.terraform.io/docs/providers/azurerm/r/storage_account.html#tags) on the parent resource `azurerm_storage_account`, but I would prefer using the `metadata` attribute.

**Comment by rrmistry** [comment]
CC: @Lucretius, @katbyte,

**Comment by rrmistry** [comment]
Thanks @tombuildsstuff ! I verified that upper case `metadata` tags **will not work** as per your comment:
https://github.com/terraform-providers/terraform-provider-azurerm/commit/1079afa4a9889304cce16d9807d8f22b56309a04#r36216441

Even with the perpetual diff I found that changing `metadata` tag does not update the underlying `azurerm_storage_blob` resource. Only updates the `metadata` tag (which is fine).

I found that adding a portion of the file hash to the resource name does refresh the underlying `azurerm_storage_blob` resource, which meets my needs.

I'll close this issue now.
